### PR TITLE
ENH: NEP index builder

### PR DESCRIPTION
**Pull request by stefanv**
Build a numbered index for the NEP documentation.  Also perform consistency checking on NEP status flags.

### CODE REVIEW

File: doc/neps/tools/build_index.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+def nep_metadata():
+    ignore = ('nep-template.rst')
+    sources = sorted(glob.glob(r'nep-*.rst'))
+    sources = [s for s in sources if not s in ignore]
+
+    meta_re = r':([a-zA-Z\-]*): (.*)'
+
+    neps = {}
+    print('Loading metadata for:')
```
Comment (eric-wieser) [inline doc/neps/tools/build_index.py]: Seems like we should be able to parse the `rst` file with `docutils`

File: doc/neps/tools/build_index.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if tags['Status'] == 'Superseded':
+            if not 'Replaced-By' in tags:
+                raise RuntimeError(
+                    f'NEP {nr} has been Superseded, but has no Replaced-By tag'
+                )
+
+            replaced_by = int(tags['Replaced-By'])
+            replacement_nep = neps[replaced_by]
+
+            if not int(replacement_nep['Replaces']) == nr:
```
Comment (eric-wieser) [inline doc/neps/tools/build_index.py]: Also needs a check that `'Replaces' in replacement_nep`

File: doc/neps/index.rst.tmpl [hunk 3]
```diff
@@ -14,43 +14,38 @@ Meta-NEPs (NEPs about NEPs or Processes)
...
-   nep-0002-warnfix
-   nep-0003-math_config_clean
-   nep-0004-datetime-proposal3
-   nep-0006-newbugtracker
-   nep-0008-groupby_additions
-   nep-0009-structured_array_extensions
-   nep-0011-deferred-ufunc-evaluation
-   nep-0012-missing-data
+{% for nep, tags in neps.items() if tags['Status'] == 'Deferred' %}
+   NEP {{ nep }} — {{ tags['Title'] }} <{{ tags['Filename'] }}>
```
Comment (eric-wieser) [inline doc/neps/index.rst.tmpl]: Possibly worth sorting the NEPs?

**Comment by stefanv** [comment]
They are sorted in the index builder.

On March 21, 2018 17:39:29 Eric Wieser <notifications@github.com> wrote:
@eric-wieser commented on this pull request.
In doc/neps/index.rst.tmpl:
> Defunct NEPs ------------ .. toctree:: :maxdepth: 1 - nep-0002-warnfix
- nep-0003-math_config_clean
- nep-0004-datetime-proposal3
- nep-0006-newbugtracker
- nep-0008-groupby_additions
- nep-0009-structured_array_extensions
- nep-0011-deferred-ufunc-evaluation
- nep-0012-missing-data
+{% for nep, tags in neps.items() if tags['Status'] == 'Deferred' %}
+ NEP {{ nep }} — {{ tags['Title'] }} <{{ tags['Filename'] }}>
Possibly worth sorting the NEPs?

**Comment by eric-wieser** [comment]
Oh, right - this is python 3.6 so dictionary items are sorted in insertion order.

File: doc/neps/nep-0007-datetime-proposal.rst [hunk 4]
```diff
@@ -14,6 +14,7 @@ Revised only slightly from the third proposal by
 :Contact: ivan@selidor.net
 :Date: 2008-07-30
 
+:Status: Final
```
Comment (njsmith) [inline doc/neps/nep-0007-datetime-proposal.rst]: Maybe this should go up at the top? This is a weird one b/c the author/contact stuff just above is referring to another NEP...

File: doc/neps/tools/build_index.py (hunk 1 above)
Comment (stefanv) [reply doc/neps/tools/build_index.py]: Indeed, it is possible, but would look  something like this: https://stackoverflow.com/questions/12883428/how-to-parse-restructuredtext-in-python

**Comment by stefanv** [comment]
@charris @eric-wieser Updated including your feedback (except for the docutils parsing, which we can perhaps leave for another time?).

**Comment by njsmith** [comment]
+1 to **not** using docutils here... this format is nice because it renders well in Github's renderer, but I'm not even sure that docutils does anything particularly special when parsing it, and doing anything with docutils is always incredibly complicated.

**Comment by eric-wieser** [comment]
The format of the `.rst` files is already docutils (it renders with sphinx), but you've ended up writing your own parser.

It's surprisingly easy to parse this:

```python
def parse_from_string(s):
    doctree = docutils.core.publish_doctree(s)

    header_node = first(doctree, lambda x: isinstance(x, nodes.title))
    header = header_node.astext()

    metadata_node = first(doctree, lambda x: isinstance(x, nodes.docinfo))
    metadata = {}

    for field_node in metadata_node:
        if isinstance(field_node, nodes.field):
            name = field_node[0].astext()
            value = field_node[1].astext()
            metadata[name] = value

    return header, metadata
```

**Comment by stefanv** [comment]
If, by parser, you mean `lines[1]`! Is the additional complexity worth it?

**Comment by stefanv** [comment]
Ah, I guess you mean the header tags.

**Comment by eric-wieser** [comment]
Yeah, I mean the header tags, which `docutils` makes easy to parse, and is also guaranteed to do so in the same way as it does in sphinx.

I'm not suggesting we generate the index with docutils (yet)

**Comment by eric-wieser** [comment]
Got my comment wrong last time - updated.

**Comment by stefanv** [comment]
Is the docutils change required for merge, or could we consider that a future improvement?

**Comment by eric-wieser** [comment]
I suppose it could be a future improvement

**Comment by stefanv** [comment]
@charris If you have a moment, could you take a look at this PR?

**Comment by charris** [comment]
I count this work as a sort of sub-project within numpy, so I expect most to the review to be done by others involved.

Thanks @stefanv.

**Comment by stefanv** [comment]
Thanks, Chuck, a little typo snuck through, fixed here: https://github.com/numpy/numpy/pull/10836

**Pull request merged**